File: solutions/Arrays & Hashing/271_encode-and-decode-strings.py
from typing import List

class Solution:

    def encode(self, strs: List[str]) -> str:
        res = ""
        for s in strs:
            res += str(len(s)) + "-" + s
        return res

    def decode(self, s: str) -> List[str]:
        res = []
        i = 0
        while i < len(s):
            curr = ""
            d = ""
            while s[i] != "-":
                d += s[i]
                i += 1
            d = int(d)

            for _ in range(d):
                i += 1
                curr += s[i]

            i += 1
            res.append(curr)

        return res
    
# Time Complexity: O(n)
# Space Complexity: O(n)

File: C++/Arrays & Hashing/271_encode-and-decode-strings.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    string encode(vector<string>& strs) {
        string res;
        for (string& s : strs) {
            res += to_string(s.size()) + "-" + s;
        }
        return res;
    }

    vector<string> decode(string s) {
        vector<string> res;
        int i = 0;

        while (i < s.size()) {
            string d;

            while (s[i] != '-') {
                d += s[i];
                i++;
            }

            int len = stoi(d);
            string curr;

            for (int k = 0; k < len; k++) {
                i++;
                curr += s[i];
            }

            i++;
            res.push_back(curr);
        }

        return res;
    }
};
